pkg/platform/runtime/alternative.go

```diff
@@ -95,32 +101,53 @@ func (ar *AlternativeRuntime) BuildEngine() BuildEngine {
 	return Alternative
 }
 
-func (ar *AlternativeRuntime) cachedArtifact(downloadDir string) *string {
-	candidate := filepath.Join(downloadDir, constants.ArtifactArchiveName)
-	if fileutils.FileExists(candidate) {
-		return &candidate
+func (ar *AlternativeRuntime) artifactCache() ([]artifactCacheMeta, error) {
+	cache := []artifactCacheMeta{}
+	jsonFile := filepath.Join(ar.runtimeEnvBaseDir(), constants.ArtifactCacheFileName)
+	if !fileutils.FileExists(jsonFile) {
+		return cache, nil
 	}
 
+	jsonBlob, fail := fileutils.ReadFile(jsonFile)
+	if fail != nil {
+		return cache, errs.Wrap(fail, "Could not read artifact cache file")
+	}
+	if err := json.Unmarshal(jsonBlob, &cache); err != nil {
+		return cache, errs.Wrap(fail, "Could not unmarshal artifact cache file")
+	}
+
+	return cache, nil
+}
+
+func (ar *AlternativeRuntime) storeArtifactCache() error {
+	// Save artifact cache information
+	jsonBlob, err := json.Marshal(ar.cache)
+	if err != nil {
+		return errs.Wrap(err, "Failed to marshal artifact cache information")
+	}
+	jsonFile := filepath.Join(ar.runtimeEnvBaseDir(), constants.ArtifactCacheFileName)
+	if fail := fileutils.WriteFile(jsonFile, jsonBlob); fail != nil {
+		return errs.Wrap(fail, "Failed to write artifact cache information")
+	}
 	return nil
 }
 
-// ArtifactsToDownloadAndUnpack returns the artifacts that we need to download
+// ArtifactsToDownload returns the artifacts that we need to download
+// for this project.
+func (ar *AlternativeRuntime) ArtifactsToDownload() []*HeadChefArtifact {
+	return artifactsToDownload(artifactCacheToUuids(ar.cache), ar.artifactsRequested)
+}
+
+// ArtifactsToDownload returns the artifacts that we need to download
 // for this project.
-// Otherwise: It filters out artifacts that have been downloaded before, and adds them to
-// the list of artifacts that need to be unpacked only.
-func (ar *AlternativeRuntime) ArtifactsToDownloadAndUnpack() ([]*HeadChefArtifact, map[string]*HeadChefArtifact) {
+func artifactsToDownload(artifactCacheUuids []strfmt.UUID, artifactsRequested []*HeadChefArtifact) []*HeadChefArtifact {
 	downloadArtfs := []*HeadChefArtifact{}
-	archives := map[string]*HeadChefArtifact{}
-
-	for downloadDir, artf := range ar.artifactMap {
-		cached := ar.cachedArtifact(downloadDir)
-		if cached == nil {
-			downloadArtfs = append(downloadArtfs, artf)
-		} else {
-			archives[*cached] = artf
+	for _, v := range artifactsRequested {
+		if v.ArtifactID != nil && !funk.Contains(artifactCacheUuids, *v.ArtifactID) {
+			downloadArtfs = append(downloadArtfs, v)
 		}
 	}
-	return downloadArtfs, archives
+	return downloadArtfs
 }
 
 // IsInstalled checks if the merged runtime environment definition file exists
```

Comment:
I think that this function will also need to be adapted.  It probably needs to check if the requested artifacts are equal to the installed artifacts.